ffi/src/schema_visitor.rs

```diff
@@ -0,0 +1,1112 @@
+/// the array can be null if and only if the field referenced by `element_type_id` is nullable.
+///
+/// # Safety
+///
+/// Caller is responsible for providing valid `state`, `name` slice, `element_type_id` from
+/// previous `visit_data_type_*` call, and `allocate_error` function pointer.
+#[no_mangle]
+pub unsafe extern "C" fn visit_field_array(
+    state: &mut KernelSchemaVisitorState,
+    name: KernelStringSlice,
+    element_type_id: usize,
+    nullable: bool,
+    allocate_error: AllocateErrorFn,
+) -> ExternResult<usize> {
+    let name_str = unsafe { TryFromStringSlice::try_from_slice(&name) };
+    visit_field_array_impl(state, name_str, element_type_id, nullable)
+        .into_extern_result(&allocate_error)
+}
+
+fn visit_field_array_impl(
+    state: &mut KernelSchemaVisitorState,
+    name: DeltaResult<&str>,
+    element_type_id: usize,
+    nullable: bool,
+) -> DeltaResult<usize> {
+    let name_str = name?.to_string();
+    let element_field = unwrap_field(state, element_type_id).ok_or_else(|| {
+        Error::generic(format!(
+            "Invalid element type ID {element_type_id} for array"
+        ))
+    })?;
+
+    let array_type = ArrayType::new(element_field.data_type, element_field.nullable);
+    let field = StructField::new(name_str, array_type, nullable);
+    Ok(wrap_field(state, field))
+}
+
+/// Visit a map field. Map fields store key-value pairs where all keys have the same type and all
+/// values have the same type.
+///
+/// Both `key_type_id` and `value_type_id` must reference fields created by previous `visit_field_*`
+/// calls. The map can contain null values if and only if the field referenced by `value_type_id` is
+/// nullable.
+///
+/// # Safety
+///
+/// Caller is responsible for providing valid `state`, `name` slice, `key_type_id` and `value_type_id`
+/// from previous `visit_data_type_*` calls, and `allocate_error` function pointer.
+#[no_mangle]
+pub unsafe extern "C" fn visit_field_map(
+    state: &mut KernelSchemaVisitorState,
+    name: KernelStringSlice,
+    key_type_id: usize,
+    value_type_id: usize,
+    nullable: bool,
+    allocate_error: AllocateErrorFn,
+) -> ExternResult<usize> {
+    let name_str = unsafe { TryFromStringSlice::try_from_slice(&name) };
+    visit_field_map_impl(state, name_str, key_type_id, value_type_id, nullable)
+        .into_extern_result(&allocate_error)
+}
+
+fn visit_field_map_impl(
+    state: &mut KernelSchemaVisitorState,
+    name: DeltaResult<&str>,
+    key_type_id: usize,
+    value_type_id: usize,
+    nullable: bool,
+) -> DeltaResult<usize> {
+    let name_str = name?.to_string();
+
+    let key_field = unwrap_field(state, key_type_id)
+        .ok_or_else(|| Error::generic(format!("Invalid key type ID {key_type_id} for map")))?;
+
+    let value_field = unwrap_field(state, value_type_id)
+        .ok_or_else(|| Error::generic(format!("Invalid value type ID {value_type_id} for map")))?;
+
```

Comment:
Maybe check that key is not nullable? That way round trip will fail instead of getting inconsistent results.


```suggestion
require!(!key_field.nullable, Error::generic("Delta Map keys may not be nullable"));
```